### PR TITLE
NetworkVisitor

### PR DESCRIPTION
**Pull request by petersteneteg**
Core: Added support of network visitation in the form of a NetworkVisitor that visits processors and properties

### CODE REVIEW

File: include/inviwo/core/network/networkvisitor.h [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    virtual ~NetworkVisitor() = default;
+
+    virtual bool visit(Processor&) { return true; }
+    virtual bool visit(PropertyOwner&) { return true; }
+    virtual bool visit(Property&) { return true; }
+    virtual bool visit(CompositeProperty&) { return true; }
+};
+
+template <typename... Funcs>
+struct LambdaNetworkVisitor : NetworkVisitor, Funcs... {
```
Comment (ResearchDaniel) [inline include/inviwo/core/network/networkvisitor.h:95]: This sounds really convenient, but I have no idea how to use it :D
A hint of an example or short description would be helpful

File: include/inviwo/core/network/networkvisitor.h [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+class CompositeProperty;
+
+/**
+ * \brief A vistor base for visiting an inviwo network
+ */
+class IVW_CORE_API NetworkVisitor {
+public:
+    virtual ~NetworkVisitor() = default;
+
+    virtual bool visit(Processor&) { return true; }
```
Comment (ResearchDaniel) [inline include/inviwo/core/network/networkvisitor.h:55]: Specify what happens when returning true/false

File: include/inviwo/core/network/processornetwork.h [hunk 3]
```diff
@@ -249,6 +250,8 @@ class IVW_CORE_API ProcessorNetwork : public Serializable,
      */
     void clear();
 
+    void accept(NetworkVisitor& visor);
```
Comment (ResearchDaniel) [inline include/inviwo/core/network/processornetwork.h:257]: Should add some details here about the order. Both in terms of network topology and also the order of the visit functions ( first processor?) 
What happens with the classes deriving from multiple ones, e.g Processor is a PropertyOwner
A short description or link to the visitor pattern would also help newcomers

File: include/inviwo/core/network/networkvisitor.h [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+namespace inviwo {
+
+class Processor;
+class PropertyOwner;
+class Property;
+class CompositeProperty;
+
+/**
+ * \brief A vistor base for visiting an inviwo network
```
Comment (ResearchDaniel) [inline include/inviwo/core/network/networkvisitor.h]: Visitor pattern base for...

**Review by ResearchDaniel — CHANGES_REQUESTED** [review]
Looks good, but I think that each accept method could explain what it is doing. 
Also, the behavior with respect to for example Processor/PropertyOwner is not clear. Will for example PropertyOwner ever be called if I visit a network or is it just Processor/Property? Same goes for CompositeProperty I guess

File: include/inviwo/core/network/networkvisitor.h (hunk 1 above)
Comment (petersteneteg) [reply include/inviwo/core/network/networkvisitor.h:95]: added an example

File: include/inviwo/core/network/networkvisitor.h (hunk 2 above)
Comment (petersteneteg) [reply include/inviwo/core/network/networkvisitor.h:55]: done

File: include/inviwo/core/network/processornetwork.h (hunk 3 above)
Comment (petersteneteg) [reply include/inviwo/core/network/processornetwork.h:257]: generally for a visitor pattern you will always get the most the overload for the most derived type.

Comment (petersteneteg) [reply include/inviwo/core/network/processornetwork.h:257]: The PropertyOwner would not be used in most scenarios, but wanted it there for completeness.

Comment (ResearchDaniel) [reply include/inviwo/core/network/processornetwork.h:257]: Come to think of it we probably also want to state if adding/removing elements (e.g. properties) are allowed during traversal.

File: include/inviwo/core/properties/propertyowner.h [hunk 5]
```diff
@@ -163,18 +161,22 @@ class IVW_CORE_API PropertyOwner : public PropertyOwnerObservable,
...
-                                   subProperties.end());
+
+    for (auto* property : properties_) {
+        if (auto p = dynamic_cast<T*>(property)) {
+            foundProperties.push_back(p);
+        }
+    }
+
+    if (recursiveSearch) {
+        for (auto* copy : compositeProperties_) {
```
Comment (ResearchDaniel) [inline include/inviwo/core/properties/propertyowner.h:174]: Copy?

File: src/core/properties/propertyowner.cpp [hunk 6]
```diff
@@ -170,35 +171,33 @@ std::vector<Property*> PropertyOwner::getPropertiesRecursive() const {
...
-            if (comp) {
-                property = comp->getPropertyByIdentifier(path[i]);
-                if (!property) return nullptr;
-            } else {
-                return nullptr;
-            }
-            ++i;
+    if (path.empty()) return nullptr;
+
+    auto last = --path.end();
```
Comment (ResearchDaniel) [inline src/core/properties/propertyowner.cpp:188]: path.back()
Ah, you mean lastIt

**Review by ResearchDaniel — APPROVED** [review]
I’m not sure if the CanvasProcessor should have special treatment in the visitor. convenient, yes, but Once you go down that road: what about other types of processors?

One would instead have to check that type in my opinion. Makes the interface a bit cleaner

**Comment by petersteneteg** [comment]
see #952

**Pull request closed without merging**